### PR TITLE
ARROW-2863: [Python] Add context manager APIs to RecordBatch*Writer/Reader classes

### PR DESCRIPTION
**Pull request by kszucs**

### CODE REVIEW

File: python/pyarrow/ipc.py [hunk 1]
```diff
@@ -56,9 +56,16 @@ class RecordBatchStreamReader(lib._RecordBatchStreamReader, _ReadPandasOption):
...
     """
+
     def __init__(self, source):
         self._open(source)
 
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        pass
```
Comment (kszucs) [inline python/pyarrow/ipc.py]: Seems like we don't have a close method on the readers, but I've added it for convenience.

**Review by pitrou — APPROVED** [review]
I'm curious, why didn't you add them to the Cython classes instead? Otherwise, looks good to me.

**Comment by kszucs** [comment]
Well this change is purely python, and only affects the user facing API.

**Comment by pitrou** [comment]
Yeah. I just wonder if other Python classes may at some point subclass the Cython classes, in which case they could the context manager automatically too.

**Comment by kszucs** [comment]
@pitrou `_open()` is called from the python side constructors, so if we move the context manager to the cython classes we cannot be sure whether the objects is opened, or we should open it from `__enter__`.

**Comment by pitrou** [comment]
No problem :-)

**Review by jorisvandenbossche — APPROVED** [review]
Looks good!

**Comment by wesm** [comment]
The test failures seem legitimate, not sure at a glance what's wrong...

**Comment by jorisvandenbossche** [comment]
The failures seem to be caused by a hanging flight test: `test_do_put_independent_read_write` doesn't finish within 10 or 20 min (and it's the same on travis, appveyor and ursabot)

Looking in that test, it is using a FlightStreamWriter object as a context manager, which subclasses `_CRecordBatchWriter`, and that base class is the one where you removed the `__enter__` and `__exit__` methods.

**Comment by kszucs** [comment]
@pitrou it is ready to go

**Review by kszucs — COMMENTED** [review]
+1

**Pull request closed without merging**